### PR TITLE
#99 'runTest' not discovered

### PR DESCRIPTION
**Pull request by crwilcox**
@hawkerm discovered that we do not match the python docs.  I am
committing his fix on his behalf.  I also have altered our tests to
reflect the change.

### CODE REVIEW

File: Python/Product/TestAdapter/TestAnalyzer.cs [hunk 1]
```diff
@@ -162,7 +162,7 @@ LocationInfo sourceLocation
             AnalysisValue classValue
         ) {
             return classValue.GetAllMembers(entry.Analysis.InterpreterContext)
-                .Where(v => v.Key.StartsWith("test"))
+                .Where(v => v.Key.StartsWith("test") || v.Key.StartsWith("runTest"))
```
Comment (int19h) [inline Python/Product/TestAdapter/TestAnalyzer.cs]: `StartsWith` is not correct for `runTest` - it must override the method, so it has to use that name exactly, unlike `test_...` pattern.

**Comment by crwilcox** [comment]
FYI:  Wait to merge until @zooba and @hawkerm can look at the changes.

**Comment by zooba** [comment]
If `runTest` is overridden then we should display the class name as the test and pass the class name to the test executor rather than `class.runTest`. From a quick reading of how `unittest` works, `runTest` is not detected as a test method, but having one there may suppress the usual method detection.

Be very sure that the behaviour is consistent between what we show, what we execute/debug, and what the basic command line tools execute (for at least Python 2.7 and 3.2-3.4, 2.6 if possible). I haven't had a chance to test it all, but I'm very sceptical that treating `runTest` as if it were spelled `testrunTest` is the right approach.

**Comment by int19h** [comment]
I experimented with it a little bit from command line, and so long as you use the fully qualified method name to run things (which I believe we do in all cases), `runTest` is no different from any other method. Indeed, you can run any arbitrary method that way, regardless of its name.

When command line is used to run the entire test case class, then the presence of any `test...` method overrides `runTest`, and the latter isn't run at all. So when generating the list of test methods, we might also want to use the same either-or approach.

**Comment by crwilcox** [comment]
My concern is that not discovering it will cause people to ask 'why is runTest not being discovered'.  That being said we should replicate the same as the commandline tools. I will make the changes to make this discover contingent on the discovery of test* tests in the class.

File: Python/Product/TestAdapter/TestAnalyzer.cs [hunk 2]
```diff
@@ -157,13 +157,28 @@ LocationInfo sourceLocation
...
         private static IEnumerable<KeyValuePair<string, IAnalysisSet>> GetTestCaseMembers(
             IPythonProjectEntry entry,
             AnalysisValue classValue
         ) {
-            return classValue.GetAllMembers(entry.Analysis.InterpreterContext)
+            var tests = classValue.GetAllMembers(entry.Analysis.InterpreterContext)
                 .Where(v => v.Key.StartsWith("test"))
                 .Where(v => v.Value.Any(m => m.MemberType == PythonMemberType.Function || m.MemberType == PythonMemberType.Method));
+
+            var runTest = classValue.GetAllMembers(entry.Analysis.InterpreterContext)
```
Comment (int19h) [inline Python/Product/TestAdapter/TestAnalyzer.cs]: You're calling GetAllMembers twice (and that call won't be lazy, unlike the rest of the query) - might want to reuse that instead.

File: Python/Product/TestAdapter/TestAnalyzer.cs [hunk 3]
```diff
@@ -157,13 +157,28 @@ LocationInfo sourceLocation
...
+                .Where(v => v.Key.Equals("runTest"))
+                .Where(v => v.Value.Any(m => m.MemberType == PythonMemberType.Function || m.MemberType == PythonMemberType.Method));
```
Comment (int19h) [inline Python/Product/TestAdapter/TestAnalyzer.cs]: This is also repeated now. Query for member type first and stash that away, then do name queries on top of that.

Comment (crwilcox) [inline Python/Product/TestAdapter/TestAnalyzer.cs]: Good point.  I figured the enumeration being delayed that wouldn't make a big difference but I can clean that up a bit.

**Comment by hawkerm** [comment]
Awesome, thanks guys! This looks great. :)

File: Python/Product/TestAdapter/TestAnalyzer.cs [hunk 4]
```diff
@@ -157,13 +157,26 @@ LocationInfo sourceLocation
...
+            var methodFunctions = classValue.GetAllMembers(entry.Analysis.InterpreterContext)
                 .Where(v => v.Value.Any(m => m.MemberType == PythonMemberType.Function || m.MemberType == PythonMemberType.Method));
+
+            var tests = methodFunctions.Where(v => v.Key.StartsWith("test"));
+            var runTest = methodFunctions.Where(v => v.Key.Equals("runTest"));
+
+            if (tests.Any()) {
+                return tests;
+            } else {
+                return runTest;
```
Comment (zooba) [inline Python/Product/TestAdapter/TestAnalyzer.cs]: In this case, I'd rather return the class object (if we can get its name) than the runTest method. That way you won't see `runTest` in the list of tests - you'll see the test itself (which is an instance of the class, if you've overridden runTest).

**Comment by crwilcox** [comment]
Changes have been made to make the 'runTest' test display by the class name in test explorer which should be more useful.

**Comment by zooba** [comment]
+1

File: Python/Product/TestAdapter/TestAnalyzer.cs [hunk 5]
```diff
@@ -157,13 +160,26 @@ LocationInfo sourceLocation
...
         private static IEnumerable<KeyValuePair<string, IAnalysisSet>> GetTestCaseMembers(
             IPythonProjectEntry entry,
             AnalysisValue classValue
         ) {
-            return classValue.GetAllMembers(entry.Analysis.InterpreterContext)
-                .Where(v => v.Key.StartsWith("test"))
+            var methodFunctions = classValue.GetAllMembers(entry.Analysis.InterpreterContext)
                 .Where(v => v.Value.Any(m => m.MemberType == PythonMemberType.Function || m.MemberType == PythonMemberType.Method));
+
+            var tests = methodFunctions.Where(v => v.Key.StartsWith("test"));
```
Comment (gilbertw) [inline Python/Product/TestAdapter/TestAnalyzer.cs]: Is it illegal to have both?  What happens if we find something with testSupport but they are using runTest

Comment (crwilcox) [inline Python/Product/TestAdapter/TestAnalyzer.cs]: it is an either or.  test* means runTest won't be run.

**Pull request merged**